webapp/packages/core-blocks/src/FormControls/useInputAutocomplete.ts

```diff
@@ -0,0 +1,154 @@
+/*
+ * CloudBeaver - Cloud Database Manager
+ * Copyright (C) 2020-2024 DBeaver Corp and others
+ *
+ * Licensed under the Apache License, Version 2.0.
+ * you may not use this file except in compliance with the License.
+ */
+import { action, computed, observable } from 'mobx';
+import { type RefObject, useEffect } from 'react';
+
+import { debounce, isFuzzySearchable, isNotNullDefined } from '@cloudbeaver/core-utils';
+
+import { useObservableRef } from '../useObservableRef.js';
+
+export type InputAutocompleteStrategy = 'startsWith' | 'contains' | 'fuzzy';
+
+interface InputAutocompleteOptions {
+  sourceHints: InputAutocompleteProposal[];
+  matchStrategy?: InputAutocompleteStrategy;
+  filter?: (suggestion: InputAutocompleteProposal, lastWord?: string) => boolean;
+}
+
+export interface InputAutocompleteProposal {
+  displayString: string;
+  replacementString: string;
+  icon?: string;
+  title?: string;
+  score?: number;
+}
+
+const INPUT_DELAY = 300;
+
+export const useInputAutocomplete = (
+  inputRef: RefObject<HTMLInputElement | HTMLTextAreaElement>,
+  { sourceHints, matchStrategy = 'startsWith', filter }: InputAutocompleteOptions,
+) => {
+  const state = useObservableRef(
+    () => ({
+      input: inputRef.current?.value as string | undefined,
+      selectionStart: inputRef.current?.selectionStart ?? null,
+      selectionEnd: inputRef.current?.value?.length ?? null,
+      replaceCurrentWord(replacement: string) {
+        const input = this.inputRef.current;
+
+        if (!input) {
+          return;
+        }
+
+        const cursorPosition = this.selectionStart;
+        const words = this.input?.split(' ');
+
+        if (!isNotNullDefined(words) || !isNotNullDefined(cursorPosition) || !isNotNullDefined(this.currentWord)) {
```

Comment:
Do we need to check words and cursor here if currentWord getter should return falsy value if there is no words or cursorPosition?